crates/component/src/wrappers/http.rs

```diff
@@ -0,0 +1,422 @@
+    )>,
+    pub(crate) headers: HeaderMap,
+    // TODO(followup): Add trailers
+}
+
+impl ResponseBuilder {
+    /// Return a new [`ResponseBuilder`] with the provided status code and body
+    pub fn new(status_code: u16, body: impl AsRef<[u8]>) -> Self {
+        Self {
+            status_code: Some(status_code),
+            body: Some(body.as_ref().to_vec()),
+            body_stream: None,
+            headers: HeaderMap::new(),
+        }
+    }
+
+    /// Return a new `ResponseBuilder` with the provided status code and body stream.
+    ///
+    /// If the provided stream is from an incoming HTTP body (either a request body or a response body),
+    /// you should provide the handle to the incoming body as well. This will ensure that the incoming
+    /// body is properly finished when the response is finished.
+    ///
+    /// ```rust
+    /// // Example that streams the incoming body to the outgoing body
+    /// use wasi::http::types::ErrorCode;
+    /// use wasmcloud_component::http::{HttpServer, Request, ResponseBuilder};
+    /// struct Component;
+    /// wasmcloud_component::http::export!(Component);
+    ///
+    /// impl HttpServer for Component {
+    ///     fn handle(request: Request) -> Result<ResponseBuilder, ErrorCode> {
+    ///         let (stream, body) = request.into_body_stream()?;
+    ///         Ok(ResponseBuilder::new_stream(200, stream, Some(body)))
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// ```rust
+    /// // Example that streams a response body to the client
+    /// use wasi::http::types::ErrorCode;
+    /// use wasmcloud_component::http::{HttpServer, Request, ResponseBuilder};
+    /// use reqwest_wasmcloud as reqwest;
+    ///
+    /// struct Component;
+    /// wasmcloud_component::http::export!(Component);
+    ///
+    /// impl HttpServer for Component {
+    ///     fn handle(_request: Request) -> Result<ResponseBuilder, ErrorCode> {
+    ///         let mut response: reqwest::Response = reqwest::get("https://example.com").map_err(|e| {
+    ///             ErrorCode::InternalError(Some(format!("failed to send outbound request {e:?}")))
+    ///         })?;
+    ///         let (stream, body) = response.bytes_stream().map_err(|e| {
+    ///             ErrorCode::InternalError(Some(format!("failed to read response body {e:?}")))
+    ///         })?;
+    ///         Ok(ResponseBuilder::new_stream(200, stream, Some(body)))
+    ///     }
+    /// }
+    /// ```
+    pub fn new_stream(
+        status_code: u16,
+        stream: wasi::io::streams::InputStream,
+        body: Option<wasi::http::types::IncomingBody>,
+    ) -> Self {
+        Self {
+            status_code: Some(status_code),
+            body: None,
+            body_stream: Some((stream, body)),
+            headers: HeaderMap::new(),
+        }
+    }
+
+    /// Helper method to return a new [`ResponseBuilder`] with a 200 status code and the provided body.
+    pub fn ok(body: impl AsRef<[u8]>) -> Self {
```

Comment:
```suggestion
    pub fn ok(body: Vec<u8>) -> Self {
```

Since this will be turned into a vector anyway, let's not force a double allocation here if the users already have a vector